File: repos/backend/internal/repositories/equipment.go
package repositories

import (
	"encoding/json"
	"errors"

	"github.com/EquipQR/equipqr/backend/internal/database"
	"github.com/EquipQR/equipqr/backend/internal/database/models"
	"github.com/google/uuid"
	"gorm.io/datatypes"
)

func GetEquipmentByID(id string) (*models.Equipment, error) {
	var eq models.Equipment
	result := database.DB.Preload("Business").First(&eq, "id = ?", id)
	if result.Error != nil {
		return nil, result.Error
	}
	return &eq, nil
}

func GetIssuesByEquipmentID(id string) ([]models.Issue, error) {
	var issues []models.Issue
	result := database.DB.Where("equipment_id = ?", id).Find(&issues)
	return issues, result.Error
}

func CreateEquipmentEntry(reqID string, status string, eqType string, location string, moreFields map[string]any) (*models.Equipment, error) {
	businessID, err := uuid.Parse(reqID)
	if err != nil {
		return nil, errors.New("invalid business_id")
	}

	_, err = GetBusinessByID(reqID)
	if err != nil {
		return nil, errors.New("business not found")
	}

	moreFieldsJSON, err := json.Marshal(moreFields)
	if err != nil {
		return nil, errors.New("invalid more_fields format")
	}

	equipment := models.Equipment{
		BusinessID: businessID,
		Status:     status,
		Type:       eqType,
		Location:   location,
		MoreFields: datatypes.JSON(moreFieldsJSON),
	}

	if err := database.DB.Create(&equipment).Error; err != nil {
		return nil, err
	}

	return GetEquipmentByID(equipment.ID.String())
}

File: repos/frontend/src/lib/api/equipment.ts
import type { Equipment } from "$lib/types/equipment";
import type { Issue } from "$lib/types/issue";

export async function getEquipmentById(id: string): Promise<Equipment | null> {
  const res = await fetch(`/api/equipment/${id}`);
  if (!res.ok) {
    console.error(`Failed to fetch equipment with id ${id}`);
    return null;
  }
  return res.json();
}

export async function getEquipmentIssuesById(
  id: string
): Promise<Issue[] | null> {
  const res = await fetch(`/api/equipment/${id}/issues`);
  if (!res.ok) {
    console.error(`Failed to fetch issues for equipment with id ${id}`);
    return null;
  }
  return res.json();
}
